tools/online_editor/src/lsp_integration.ts

```diff
@@ -1,27 +1,84 @@
 // Copyright © SixtyFPS GmbH <info@slint-ui.com>
 // SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-commercial
 
+export {
+    Position as LspPosition,
+    Range as LspRange,
+} from "vscode-languageserver-types";
 import {
     Position as LspPosition,
     Range as LspRange,
 } from "vscode-languageserver-types";
+
+import { TextRange, TextPosition } from "./text";
+
 import * as monaco from "monaco-editor/esm/vs/editor/editor.api";
 
 function find_model(
-    model: string | monaco.editor.ITextModel,
+    model: string | monaco.editor.ITextModel | null | undefined,
 ): monaco.editor.ITextModel | null {
     if (typeof model === "string") {
         return monaco.editor.getModel(monaco.Uri.parse(model));
     }
-    return model;
+    return model ?? null;
+}
+
+export function editor_position_to_lsp_position(
+    model_: string | monaco.editor.ITextModel | null | undefined,
+    pos: TextPosition | null | undefined,
+): LspPosition | null {
+    const model = find_model(model_);
+    if (model == null || pos == null) {
+        return null;
+    }
+
+    // LSP line numbers are zero based (https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocuments)
+    // and Monaco's line numbers start at 1 (https://microsoft.github.io/monaco-editor/api/classes/monaco.Position.html#lineNumber)
+    const lspLine = pos.lineNumber - 1;
+    // Convert lsp utf-8 character index to JavaScript utf-16 string index
+    const line = model.getLineContent(pos.lineNumber);
+    const line_utf8 = new TextEncoder().encode(line.slice(0, pos.column));
+    const lspCharacter = line_utf8.length;
+
+    return {
+        line: lspLine,
+        character: lspCharacter,
+    };
+}
+
+export function editor_range_to_lsp_range(
+    model: string | monaco.editor.ITextModel | null | undefined,
+    range: TextRange | null | undefined,
+): LspRange | null {
+    if (range == null) {
+        return null;
+    }
+
+    const start = editor_position_to_lsp_position(model, {
+        lineNumber: range.startLineNumber,
+        column: range.startColumn,
+    });
+    const end = editor_position_to_lsp_position(model, {
+        lineNumber: range.endLineNumber,
+        column: range.endColumn,
+    });
+
+    if (start == null || end == null) {
+        return null;
+    }
+
+    return {
+        start: start,
+        end: end,
+    };
 }
 
 export function lsp_position_to_editor_position(
-    model_: string | monaco.editor.ITextModel,
-    pos: LspPosition,
-): monaco.IPosition | null {
+    model_: string | monaco.editor.ITextModel | null | undefined,
```

Comment:
I'm curious about one detail: the old code didn't allow model to be null. Now it is allowed, but AFAICS it will just result in this function returning null. Wouldn't it make sense to catch this condition at the call site at compile time(via the type system), instead of at runtime?

Reply:
Right, we're missing Rusts try operator perhaps :)